dandi/upload.py

```diff
@@ -573,3 +587,272 @@ def upload_agg(*ignored):
+
+        Yields
+        ------
+        dict
+          Records for pyout
+        """
+        # Ensure consistent types
+        path = Path(path)
+        relpath = PurePosixPath(relpath)
+        try:
+            try:
+                path_stat = path.stat()
+                yield {"size": path_stat.st_size}
+            except FileNotFoundError:
+                yield skip_file("ERROR: File not found")
+                return
+            except Exception as exc:
+                # without limiting [:50] it might cause some pyout indigestion
+                yield skip_file("ERROR: %s" % str(exc)[:50])
+                return
+
+            file_metadata_ = {
+                "uploaded_size": path_stat.st_size,
+                "uploaded_mtime": ensure_strtime(path_stat.st_mtime),
+                # "uploaded_date": None,  # to be filled out upon upload completion
+            }
+
+            #
+            # Validate first, so we do not bother server at all if not kosher
+            #
+            # TODO: enable back validation of dandiset.yaml
+            if path.name != dandiset_metadata_file and validation != "skip":
+                yield {"status": "validating"}
+                validation_errors = validate_file(path)
+                yield {"errors": len(validation_errors)}
+                # TODO: split for dandi, pynwb errors
+                if validation_errors:
+                    if validation == "require":
+                        yield skip_file("failed validation")
+                        return
+                else:
+                    yield {"status": "validated"}
+            else:
+                # yielding empty causes pyout to get stuck or crash
+                # https://github.com/pyout/pyout/issues/91
+                # yield {"errors": '',}
+                pass
+
+            #
+            # Special handling for dandiset.yaml
+            # Yarik hates it but that is life for now. TODO
+            #
+            if path.name == dandiset_metadata_file:
+                # TODO This is a temporary measure to avoid breaking web UI
+                # dandiset metadata schema assumptions.  All edits should happen
+                # online.
+                yield skip_file("should be edited online")
+                return
+
+            #
+            # Extract metadata - delayed since takes time, but is done before
+            # actual upload, so we could skip if this fails
+            #
+            # Extract metadata before actual upload and skip if fails
+            # TODO: allow for for non-nwb files to skip this step
+            # ad-hoc for dandiset.yaml for now
+            yield {"status": "extracting metadata"}
+            try:
+                metadata = get_metadata(path)
+            except Exception as exc:
+                if allow_any_path:
+                    yield {"status": "failed to extract metadata"}
+                    metadata = {}
```

Comment:
How should `encodingFormat` be determined for arbitrary files?

Reply:
I think (@satra could correct me if I remember it wrong) but for starters we decided to go with the mime type, e.g.

```shell
$> python3 -c 'import magic; print(magic.from_file("18516007.nwb", mime=True))'
application/x-hdf
```

Reply:
What PyPI project is `magic` from?  Dandi-cli currently has no dependency on such a project.

Reply:
https://pypi.org/project/python-magic/

Reply:
but grr: in conda it exists only for osx and linux... and it is a "long known" to us issue: https://github.com/conda-forge/python-magic-feedstock/issues/9

<details>
<summary>it installs ok from pypi and operates nicely on windows </summary> 

```shell
(datalad-py3.7) C:\Users\DataLad\PycharmProjects\datalad>pip install python-magic
Collecting python-magic
  Downloading python_magic-0.4.18-py2.py3-none-any.whl (8.6 kB)
Installing collected packages: python-magic
Successfully installed python-magic-0.4.18

(datalad-py3.7) C:\Users\DataLad\PycharmProjects\datalad>python
Python 3.7.9 (default, Aug 31 2020, 17:10:11) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import magic
>>> magic.from_file('setup.py')
'Python script, ASCII text executable, with CRLF line terminators'
>>> magic.from_file('setup.py', mime=True)
'text/x-python'
```
</details>

so in summary: we have  a way forward, although windows installation via conda would not be complete and would require subsequent `pip install python-magic`

Reply:
I take it back: I quickly realized that I was in a datalad conda env, in which I had our build of git-annex installed, where we do provide libmagic build

<details>
<summary>but any other environment on windows would not have libmagic installed, so even if it pip installs, it would fail to operate</summary> 

```shell
(test-magic) C:\Users\DataLad\PycharmProjects\datalad>pip install python-magic
Collecting python-magic
  Using cached python_magic-0.4.18-py2.py3-none-any.whl (8.6 kB)
Installing collected packages: python-magic
Successfully installed python-magic-0.4.18
WARNING: You are using pip version 20.1.1; however, version 20.3.3 is available.
You should consider upgrading via the 'C:\Users\DataLad\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip' command.

(test-magic) C:\Users\DataLad\PycharmProjects\datalad>python
Python 3.7.9 (tags/v3.7.9:13c94747c7, Aug 17 2020, 16:30:00) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import magic
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\DataLad\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.7_qbz5n2kfra8p0\LocalCache\local-packages\Python37\site-packages\magic.py", line 201, in <module>
    raise ImportError('failed to find libmagic.  Check your installation')
ImportError: failed to find libmagic.  Check your installation
```
</details>

so let's postpone dealing with encodingFormat  until we can ensure consistent operation across OSes